File: matlab/unit/demo_oe.m
%  Running this demos PEM estimation of Output-Error model structure

clear; close all;  
global dm; if isempty(dm), clear global dm; dm=0; end
global dsp; if isempty(dsp), clear global dsp; dsp=1; end
global trans;
if dsp>1, clc; echo on; end
OPT.dsp = dsp;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify Experiment Conditions                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

T   = 1;     % Sampling period in seconds
N   = 100;   % Number of samples
var = 1e-1;  % White Measurement Noise variance

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify true (linear) system                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

den       = real(poly([-0.1,-1,-0.2,-0.3,-0.5,-0.05+j*3,-0.05-j*3]));
num       = 10*den(length(den));
[a,b,c,d] = tf2ss(num,den);
[a,b]     = c2d(a,b,T); 
[bq,aq]   = ss2tf(a,b,c,d,1);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Simulate a data record                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

t     = 0:1:N-1; 
Z.u   = sign(sin(3*pi*t/N));
noise = sqrt(var)*randn(size(Z.u));
Z.y   = filter(bq,aq,Z.u)+noise; 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify Model Structure                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Mq.A     = length(aq)-2; 
Mq.B     = Mq.A-1; 
Mq.T     = T; 
Mq.delay = 1;
Mq.type  = 'oe';

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Estimate on basis of noise corrupted data
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Gq = est(Z,Mq,OPT); 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Plot the results
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

if dsp,
	Gt.A = aq;
	Gt.B = bq;
	Gt.T = T;
    Gt.w = Gq.w;
	Gt.type='oe';
	Gt.disp.colour='b';
	Gt.disp.legend = 'True Response';
	Gq.disp.legend = [Gq.disp.legend, ' q operator'];
	showbode(Gt,Gq);
end

echo off;


if dm
 disp('  ')
 disp('---------------------------------------------------------------------')
 disp('  ')
 disp('You now have access to the MATLAB workspace so that you may examine')
 disp('the results of this simulation.  To return to the demos, type "dbcont"')
 disp(' ')
 keyboard; 
end;

File: src/uonidtoolbox/demo/_demo_oe.py

import uonidtoolbox as unit
import numpy as np
import scipy


def demo_oe(disp=1):

    OPT = unit.struct()
    OPT.dsp = disp

    # ====================================================
    # Specify Experiment Conditions
    # ====================================================
    T   = 1   # Sampling period in seconds
    N   = 500   # Number of Samples
    var = 1e-1  # Measurement Noise Variance


    # ====================================================
    # Specify true (linear) system
    # ====================================================
    den         = np.real(np.poly([-0.1,-1,-0.2,-0.3,-0.5,-0.05+1j*3,-0.05-1j*3]))
    num         = 10*den[-1]
    sysc        = scipy.signal.TransferFunction(num, den).to_ss()
    A,B,C,D,_   = scipy.signal.cont2discrete((sysc.A, sysc.B, sysc.C, sysc.D), T, method='zoh')
    bq,aq       = scipy.signal.ss2tf(A, B, C, D, 0)
    bq,aq       = np.squeeze(bq), np.squeeze(aq)


    # ====================================================
    # Simulate a data record
    # ====================================================
    Z       = unit.struct()
    Z.u     = np.sign(np.sin(3*np.pi*np.arange(0, N-1, 1)))
    noise   = np.sqrt(var)*np.random.randn(Z.u.size).reshape(Z.u.shape)
    Z.y     = scipy.signal.lfilter(bq, aq, Z.u) + noise


    # ====================================================
    # Specify Model Structures
    # ====================================================
    Mq          = unit.struct()
    Mq.A        = aq.shape[0]-1
    Mq.B        = Mq.A-1
    Mq.T        = T
    Mq.delay    = 1
    Mq.type     = 'oe'


    # ====================================================
    # Estimate on basis of noise corrupted data
    # ====================================================
    Gq = unit.est(Z,Mq,OPT)


    # ====================================================
    # Plot the results
    # ====================================================
    if OPT.dsp:
        Gt              = unit.struct()
        Gt.disp         = unit.struct()
        Gt.A            = aq
        Gt.B            = bq
        Gt.T            = T
        Gt.w            = Gq.w
        Gt.type         = 'oe'
        # Gt.colour       = 'b'
        Gt.disp.legend  = 'True Response'
        Gt.disp.legend  = Gt.disp.legend + ' q operator'
        
        unit.plotting.showbode([Gt, Gq])
    #endif

#endfunction

